### ISSUE TITLE
Parsing issue when city name is not provided and province is included in the address

### ISSUE THREAD

**Issue by Randyklose** [openvenues/libpostal#254]
The address i was testing is 322-b boul rene-levesque, a12s3d, canada.

if city isn't provided then the street name (rene-levesque) is considered as the city:

> 322-b boul rene-levesque, qc, a12s3d, canada
Result:
{
  "house_number": "322-b",
  "road": "boul",
  "city": "rene-levesque",
  "state": "qc",
  "postcode": "a12s3d",
  "country": "canada"
}

However, if the city name is included in the address then it is tokenized correctly:

> 322-b boul rene-levesque,montreal, qc, a12s3d, canada
Result:
{
  "house_number": "322-b",
  "road": "boul rene-levesque",
  "city": "montreal",
  "state": "qc",
  "postcode": "a12s3d",
  "country": "canada"
}

Just wondering how this scenario is handled in the parser. Thanks

**Comment by albarrentine** [comment]
Wrote about this in #178, but in the v1.0 release we were a bit overzealous in requiring `city` to be listed on all training examples. One aspect of the machine learning model we use, the Conditional Random Field, is that there are transition weights between each pair of labels, so the parser in this case gets very confident that if it sees `state` (and especially the token "QC" labeled as `state`), the previous token must be `city`, even if it's not a known place name.

This causes a few issues in edge case-y situations in the US and Canada, but generally helps performance in the rest of the world. There are two reasons why, one technical and one human:

1. Many major cities around the world are part of an eponymous district, state, and even country in cases like Singapore or Luxembourg. When we create the training data from OSM data, we first reverse geocode to every OSM polygon, including all administrative levels, but to format that address as a training example, we want to essentially mimic how a human would write it. So we have a set of configs which specify how likely each component should be in a given country or language. In the US, for example, we would usually see city+state, whereas state_district and country are more rare. If we never drop the `city` tag (when it's available), we can encourage the parser to prefer `city` when it sees the place name by itself. Dropping the city tag in the training data would create examples like the one above with "QC", but it would also create examples where, say, "New York" is listed in the exact same address meaning either `state` or `city` depending on a coin flip, with no real contextual way to disambiguate.

2. In some countries we don't have complete training data, partially because, as I've said in the 1.0 blog post, the mapping community is not all that representative of the diversity of the world. The question of where and what gets mapped often reflects the the people doing the mapping and their values. If we don't have every place name in some countries as part of OSM, the transition weights may still be able to help get the correct parse on structure alone.

The solution is to simply to have a very small but non-zero probability of `city` dropout in the US and Canada, which has been added for the v1.1 training data. The release is delayed a bit at present because OSM has grown too large to create the training set on a single machine, so I've had to develop some distributed Spark-based methods for point-in-polygon testing to do the reverse geocoding piece across multiple machines, and there's some refactoring associated with that as well.

This and similar parses should work well when v1.1 is out.

**Comment by Randyklose** [comment]
Great. Thanks for the quick response and sorry i missed the already existing issue. Good luck finishing the refactoring for v1.1
